### PR TITLE
Code Improvements to IDTokenBuilder and OIDC UserInfo endpoint

### PR DESCRIPTION
**Pull request by mefarazath**
### Proposed changes in this pull request
- Code improvements to IDTokenBuilder
- Code improvements to UserInfoEndpoint

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/user/OpenIDConnectUserEndpoint.java [hunk 1]
```diff
@@ -119,63 +122,63 @@ public Response getUserClaimsPost(@Context HttpServletRequest request) throws OA
...
+        return Response.status(response.getResponseStatus()).entity(response.getBody()).build();
+    }
+
+    private Response buildBadRequestErrorResponse(UserInfoEndpointException ex,
+                                                  int statusCode) throws OAuthSystemException {
+        OAuthResponse res = OAuthASResponse.errorResponse(statusCode)
+                .setError(ex.getErrorCode())
+                .setErrorDescription(ex.getErrorMessage())
+                .buildJSONMessage();
+
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/user/OpenIDConnectUserEndpoint.java]: Shall we remove this line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java [hunk 2]
```diff
@@ -81,10 +83,63 @@ protected String getSubjectClaim(Map<String, Object> userClaims,
...
+        return IdentityTenantUtil.getTenantDomain(tenantId);
+
+    }
+
+    protected Map<String, Object> getEssentialClaims(OAuth2TokenValidationResponseDTO tokenResponse,
+                                                     Map<String, Object> claims) throws UserInfoEndpointException {
+
+        Map<String, Object> essentialClaimMap = new HashMap<>();
+        List<String> essentialClaims = getEssentialClaimUris(tokenResponse);
+        if (CollectionUtils.isNotEmpty(essentialClaims)) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java]: Shall we use static import for CollectionUtils :)

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java [hunk 3]
```diff
@@ -81,10 +83,63 @@ protected String getSubjectClaim(Map<String, Object> userClaims,
...
+    }
+
+    protected abstract Map<String, Object> retrieveUserClaims(OAuth2TokenValidationResponseDTO tokenValidationResponse)
+            throws UserInfoEndpointException;
+
+    /**
+     * Build UserInfo response to be sent back to the client.
+     *
+     * @param tokenResponse      {@link OAuth2TokenValidationResponseDTO} Token Validation response containing metadata
+     *                           about the access_token used for use rinfo call.
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java]: type : use rinfo --> user info

Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/AbstractUserInfoResponseBuilder.java]: access_token -> access token

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java [hunk 4]
```diff
@@ -189,26 +183,87 @@ public String buildIDToken(OAuthTokenReqMessageContext tokenReqMsgCtxt,
...
+        JWTClaimsSet jwtClaimsSet = new JWTClaimsSet();
+        jwtClaimsSet.setIssuer(issuer);
+
+        // Set the audience
+        List<String> audience = getOIDCAudience(clientId);
+        jwtClaimsSet.setAudience(audience);
+
+        jwtClaimsSet.setClaim(AZP, clientId);
+        jwtClaimsSet.setExpirationTime(getIdTokenExpiryInMillis(idTokenLifeTimeInMillis, currentTimeInMillis));
+        jwtClaimsSet.setIssueTime(new Date(currentTimeInMillis));
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: Don't we need to use UTC time here? Calendar.getInstance(TimeZone.getTimeZone(UTC)

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java [hunk 5]
```diff
@@ -189,26 +183,87 @@ public String buildIDToken(OAuthTokenReqMessageContext tokenReqMsgCtxt,
...
+        }
+
+        String responseType = authzReqMessageContext.getAuthorizationReqDTO().getResponseType();
+        String atHash = null;
+        if (isIDTokenSigned() && isAccessTokenHashApplicable(responseType)) {
+            atHash = getAtHash(accessToken);
+        }
+
+        if (atHash != null) {
+            jwtClaimsSet.setClaim(AT_HASH, atHash);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: Shall we move this inside above if statement?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java (hunk 4 above)
Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: new Data() does that internally i guess.

**Pull request merged**